### ISSUE TITLE
False-positive PossiblyUndefinedStringArrayOffset after using `array_map`

### ISSUE THREAD

**Issue by fluffycondor** [vimeo/psalm#7946]
https://psalm.dev/r/84d60f35b3

The offset `bar` definitely exists in every subarray.

**Comment by AndrolGenhald** [comment]
Indeed, not just with `array_map` though: https://psalm.dev/r/5eedbca82f
`$baz` could be improved to be `array{foo: 'bar'|int}<array-key, int>` (which is a weird internal type definition for an array shape with extra values of known type).
